src/_posts/2020-03-22-types-and-accepting-the-facts.md

```diff
@@ -0,0 +1,525 @@
+
+```java
+class SearchService {
+
+    public List<Flight> findFlights(String startDate, String endDate, String origin, String dest) {
+        LocalDate flightStartDate = parseDate(startDate);
+        LocalDate flightEndDate = parseDate(endDate);
+        return searchRepository.findFlights(flightStartDate, flightEndDate, origin, dest);
+    }    
+}
+```
+
+```java    
+    class SearchRepository {
+    
+    		public List<Flight> findFlights(LocalDate startDate, LocalDate endDate, String origin, String dest) {
+    			// implementation
+    		}    
+    }
+```
+
+```java    
+class DateTimeFormatter {
+ 
+    private static LocalDate parseDate(String date) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+        return LocalDate.parse(date, formatter);
+    }
+
+}
+```
+
+Now with the parser being done in the service we replace exceptions related to our database for `DateTimeParseException`, this makes way easier to capture the right exception instead of trying to figure it out what was happening. What we have now is better than the previous code using strings all around but we can and must do better. The `SearchService` is throwing `DateTimeParseException` and we can handle that case in the controller and return something like `400 - Bad Request`.
+
+Now let's take care of the IATA, the IATA specification (source: Wikipedia, I didn't read the specification) says that's a code composed by three letters. In this case, we can create a class for it and add the validation. 
+
+```java
+class InvalidIATAException extends InvalidArgumentException {}
+```
+
+```java
+class IATA {
+        
+    public final String iata;
+
+    public IATA(String iata) {
+        if (iata.length() != 3) {
+            throw new InvalidIATAException();
+        }
+        this.iata = iata;
+    }
+}
+```
+
+Then we change the service and the repository to start using types:
+
+```java
+class SearchService {
+
+    public List<Flight> findFlights(String startDate, String endDate, String origin, String dest) {
+        LocalDate flightStartDate = parseDate(startDate);
+        LocalDate flightEndDate = parseDate(endDate);
+        IATA originAirport = new IATA(origin);
+        IATA destAirport = new IATA(dest);
+        return searchRepository.findFlights(flightStartDate, flightEndDate, origin, dest);
+    }
+}
+```
+
+```java
+class SearchRepository {
+
+    public List<Flight> findFlights(LocalDate startDate, LocalDate endDate, IATA origin, IATA dest) {
+        // implementation
+    }
+}
+```    
+
+With those changes, we can at least guarantee that the dates passed to the database are valid and the formatting don't matter much at this point now because it's encapsulated inside of a class. We can't confirm that the date exists in the database but there won't be any exceptions thrown when the query is executed. 
```

Comment:
With those changes, we can at least guarantee that the dates passed to the database are valid and the formatting doesn't matter much at this point now because it's encapsulated inside of a class.